### ISSUE TITLE
Epson ET-2810 IPP printing doesn't work, JOBSTATE: document-format-error

### ISSUE THREAD

**Issue by markwellis** [OpenPrinting/cups-filters#436]
I Can't print PDFs using the IPP Everywhere/driverless setup.

I Can print a test page from cups, or a document from libreoffice writer.

seems similar to #426 

Fedora 35

cups.x86_64                                    1:2.3.3op2-11.fc35
cups-client.x86_64                             1:2.3.3op2-11.fc35
cups-filesystem.noarch                         1:2.3.3op2-11.fc35
cups-filters.x86_64                            1.28.10-1.fc35    
cups-filters-libs.x86_64                       1.28.10-1.fc35    
cups-ipptool.x86_64                            1:2.3.3op2-11.fc35
cups-libs.i686                                 1:2.3.3op2-11.fc35
cups-libs.x86_64                               1:2.3.3op2-11.fc35
cups-pk-helper.x86_64                          0.2.6-13.fc35     
gutenprint-cups.x86_64                         5.3.4-5.fc35      

I've attached error logs, spool (s00041-001.gz), and source (test.pdf)
[d00041-001.gz](https://github.com/OpenPrinting/cups-filters/files/7773569/d00041-001.gz)
[error.log](https://github.com/OpenPrinting/cups-filters/files/7773570/error.log)
[test.pdf](https://github.com/OpenPrinting/cups-filters/files/7773571/test.pdf)

Thanks

**Comment by tillkamppeter** [comment]
Could you also attach the file `/etc/cups/ppd/EPSON_ET-2810_Series.ppd` (if it does not work, copy it and rename the copy to have a `.txt` extension).

Please run the following command:
```
ipptool -tv ipps://EPSON033C1E.local:631/ipp/print get-printer-attributes.test > ipp-attrs.txt
```
and attach the resulting file `ipp-attrs.txt` here.

**Comment by markwellis** [comment]
Both attached. Thanks

[EPSON_ET-2810_Series.ppd.txt](https://github.com/OpenPrinting/cups-filters/files/7804319/EPSON_ET-2810_Series.ppd.txt)
[ipp-attrs.txt](https://github.com/OpenPrinting/cups-filters/files/7804320/ipp-attrs.txt)

**Comment by tillkamppeter** [comment]
@markwellis thank you for the files.

Could you run the commands
```
sudo mv /usr/share/cups/mime/cupsfilters-ghostscript.convs /usr/share/cups/cupsfilters-ghostscript.convs
killall -HUP cupsd
```
After that print the job again. Does it work now?

**Comment by markwellis** [comment]
Didn't work

**Comment by tillkamppeter** [comment]
@markwellis could you attach an error_log of this attempt? To see whether actually `pdftoraster` instead of `gstoraster` got used and whether `pdftoraster` did the right thing?

**Comment by markwellis** [comment]
[cups_log.txt](https://github.com/OpenPrinting/cups-filters/files/7820815/cups_log.txt)

**Comment by markwellis** [comment]
anything else I can get/try to help debug it? let me know, I'm happy to try things

**Comment by TinyTrebuchet** [comment]
Can you try this command and attach the 'out' file:
`cupsfilter -d EPSON_ET-2810_Series test.pdf > out`
Also try to directly print the 'out' file:
`lp -d EPSON_ET-2810_Series -o raw out`

**Comment by markwellis** [comment]
That printed 4 pages of gibberish and I stopped it before it finished.
![image](https://user-images.githubusercontent.com/5102352/149172040-493cb579-8a7f-487d-87cb-17e084e41ba2.png)

`/usr/sbin/cupsfilter -d EPSON_ET-2810_Series test.pdf > out`
```
EBUG: argv[0]="EPSON_ET-2810_Series"
DEBUG: argv[1]="1"
DEBUG: argv[2]="mark"
DEBUG: argv[3]="test.pdf"
DEBUG: argv[4]="1"
DEBUG: argv[5]=""
DEBUG: argv[6]="test.pdf"
DEBUG: envp[0]="<CFProcessPath>"
DEBUG: envp[1]="CONTENT_TYPE=application/pdf"
DEBUG: envp[2]="CUPS_DATADIR=/usr/share/cups"
DEBUG: envp[3]="CUPS_FONTPATH=/usr/share/cups/fonts"
DEBUG: envp[4]="CUPS_SERVERBIN=/usr/lib/cups"
DEBUG: envp[5]="CUPS_SERVERROOT=/etc/cups"
DEBUG: envp[6]="LANG=en_GB.UTF8"
DEBUG: envp[7]="PATH=/usr/lib/cups/filter:/usr/bin:/usr/sbin:/bin:/usr/bin"
DEBUG: envp[8]="PPD=/usr/share/cups/model/laserjet.ppd"
DEBUG: envp[9]="PRINTER_INFO=EPSON ET-2810 Series"
DEBUG: envp[10]="PRINTER_LOCATION="
DEBUG: envp[11]="PRINTER=EPSON_ET-2810_Series"
DEBUG: envp[12]="RIP_MAX_CACHE=128m"
DEBUG: envp[13]="USER=mark"
DEBUG: envp[14]="CHARSET=utf-8"
DEBUG: envp[15]="FINAL_CONTENT_TYPE=application/pdf"
INFO: gziptoany (PID 86673) started.
INFO: gziptoany (PID 86673) exited with no errors.
```
out attached, renamed to .pdf for upload
[out.pdf](https://github.com/OpenPrinting/cups-filters/files/7855742/out.pdf)

`lp -d EPSON_ET-2810_Series -o raw out` output
```
request id is EPSON_ET-2810_Series-49 (1 file(s))
```

cups log
[cups_log.txt](https://github.com/OpenPrinting/cups-filters/files/7855800/cups_log.txt)

hopefully this is useful :)

For what it's worth, it's not an urgent problem as the escpr drivers from https://www.openprinting.org/driver/epson-escpr/ work fine, until cups 3 drops support for them

Thanks

**Comment by tillkamppeter** [comment]
@markwellis unfortunately, your CUPS error_log from your comment 6 days ago is only a tiny part of the actual log of the job. Could you print another job (PDF file, NO `-o raw`), copy the whole `error_log`, and either attach the whole file or search from the beginning of the file on for the first occurrence of `Job XX` with `XX` being the job number. Attach the `error_log` from that point on up to the end.

**Comment by tillkamppeter** [comment]
Note also that the only data format of the standard formats for driverless printing which your printer supports is PWG Raster. Especially it does not support PDF, so raw printing of PDF files will not result in anything useful.

**Comment by markwellis** [comment]
> @markwellis unfortunately, your CUPS error_log from your comment 6 days ago is only a tiny part of the actual log of the job. Could you print another job (PDF file, NO `-o raw`), copy the whole `error_log`, and either attach the whole file or search from the beginning of the file on for the first occurrence of `Job XX` with `XX` being the job number. Attach the `error_log` from that point on up to the end.

this worked. here's the log for that
[cup_log.txt](https://github.com/OpenPrinting/cups-filters/files/7856574/cup_log.txt)

here's the error log for printing via gtk print dialog in Atril
[cup_log_gtk.txt](https://github.com/OpenPrinting/cups-filters/files/7856586/cup_log_gtk.txt)






> Note also that the only data format of the standard formats for driverless printing which your printer supports is PWG Raster. Especially it does not support PDF, so raw printing of PDF files will not result in anything useful.

I'm not sure what you mean by this? Maybe I misunderstood what you meant when you said "Also try to directly print the 'out' file"?

> Can you try this command and attach the 'out' file:
> 
> `cupsfilter -d EPSON_ET-2810_Series test.pdf > out`
> 
> Also try to directly print the 'out' file:
> 
> `lp -d EPSON_ET-2810_Series -o raw out`

**Comment by tillkamppeter** [comment]
> I'm not sure what you mean by this? Maybe I misunderstood what you meant when you said "Also try to directly print the 'out' file"?
> 
> > Can you try this command and attach the 'out' file:
> > `cupsfilter -d EPSON_ET-2810_Series test.pdf > out`
> > Also try to directly print the 'out' file:
> > `lp -d EPSON_ET-2810_Series -o raw out`

This did not work because the `cupsfilter ...` command did not do the expected. `out` should be PWG Raster, but with the (wrong) `cupsfilter` command line used here `out` is PDF which the printer does not understand and therefore the `lp ... -o raw ...` prints garbage.

Try
```
cupsfilter -p EPSON_ET-2810_Series.ppd -m image/pwg-raster test.pdf > out
```
instead. Now you can display `out` with `rasterview` and should be able to print it with `lp ... -o raw ...`.

**Comment by tillkamppeter** [comment]
Thanks a lot for your great cooperation. It helped me to fix this bug.

This was only a bug in 1.x, not in master. I have fixed it now by commit 8d88225762a0 which is the same fix as commit d77db890 which I applied to master some months ago, to make the [Gutenprint Printer Application](https://snapcraft.io/gutenprint-printer-app) work correctly.

The problem was that when you printed from the command line the correct color space (#19, sRGB) was used for the PWG Raster data sent to the printer but when you printed via the GTK print dialog the job was sent with a wrong, unsupported color space (#1, RGB).

One sees this in your `error_log` files. The one of the job from the command line contains the following Ghostscript command line:
```
[Job 50] Ghostscript command line: gs -dQUIET -dSAFER -dNOPAUSE -dBATCH -dNOINTERPOLATE -dNOMEDIAATTRS -dShowAcroForm -sstdout=%stderr -sOutputFile=%stdout -sDEVICE=cups -sMediaClass=PwgRaster -sMediaType=Stationery -r360x360 -dDEVICEWIDTHPOINTS=595 -dDEVICEHEIGHTPOINTS=841 -dcupsBitsPerColor=8 -dcupsColorOrder=0 -dcupsColorSpace=19 -dcupsBorderlessScalingFactor=0.0000 -dcupsInteger1=1 -dcupsInteger2=1 -scupsPageSizeName=A4 -I/usr/share/cups/fonts -c -f -_
```
The one with the job from the GTK dialog contains:
```
[Job 51] Ghostscript command line: gs -dQUIET -dSAFER -dNOPAUSE -dBATCH -dNOINTERPOLATE -dNOMEDIAATTRS -dShowAcroForm -sstdout=%stderr -sOutputFile=%stdout -sDEVICE=cups -sMediaClass=PwgRaster -sMediaType=Stationery -r360x360 -dDEVICEWIDTHPOINTS=595 -dDEVICEHEIGHTPOINTS=841 -dcupsBitsPerColor=8 -dcupsColorOrder=0 -dcupsColorSpace=1 -dcupsBorderlessScalingFactor=0.0000 -dcupsInteger1=1 -dcupsInteger2=1 -dcupsInteger8=4 -scupsPageSizeName=iso_a4_210x297mm -I/usr/share/cups/fonts -c -f -_
```
The important difference is `-dcupsColorSpace=...` which is 19 for the command line and 1 for the GTK dialog.

To find the cause of the difference we look at the 5th command line argument which CUPS is supplying to all filters, including the `gstoraster` filter which is the one calling Ghostscript. For the command line case we have (also taken from the `error_log` files):
```
[Job 50] argv[5]="finishings=3 number-up=1 job-uuid=urn:uuid:65eb32f7-e2b4-33f5-5e5d-bacd2a05418a job-originating-host-name=localhost date-time-at-creation= date-time-at-processing= time-at-creation=1642008770 time-at-processing=1642008770 document-name-supplied=test.pdf"
```
and for the GTK dialog:
```
[Job 51] argv[5]="ColorModel=RGB OutputBin=FaceUp MediaType=Stationery number-up=1 noCollate PageSize=A4 cupsPrintQuality=Normal job-uuid=urn:uuid:99e1bee4-98ac-368a-7101-d3c09267b307 job-originating-host-name=localhost date-time-at-creation= date-time-at-processing= time-at-creation=1642008949 time-at-processing=1642008949 print-quality=4"
```
The GTK dialog adds all PPD options to the attributes of the job, regardless whether the user has changed them or not, the command line print utility `lp` only adds the options which the user explicitly specifies with `-o`. The common attributes are generated by CUPS. Especially the GTK dialog version has `ColorModel=RGB` which is not present for the command line.

So in the command line case the PPD default is used for `ColorModel` which is `RGB` and the pseudo-PostScript code for this setting in the PPD is
```
*ColorModel RGB: "<</cupsColorSpace 19/cupsBitsPerColor 8/cupsColorOrder 0/cupsCompression 0>>setpagedevice"
```
which selects color space #19. All-in-all, the PPD has only 2 `ColorModel` settings, the other is `Gray` with color space #18 (SGray), so there should never occur that #1 gets selected.

The problem is a bug in the implementation about how the `gstoraster` filter supports being used without PPD at all. For this all command line options are tried to be interpreted at best by their names andthe choices names, which makes sense when there is actually no PPD (and so no additional details about the options and choices), but the interpretation also takes place when there is a PPD and it takes place on all options/attributes in the 5th command line argument, regardless whether they are in the PPD or not.

If the options are not specified (command line case) there is no interpretation of them, and simply the PPD's default settings are used, in our case `RGB` with color space #19. If `ColorModel=19` is specified, the PPD is used at first and so also color space #19 selected but afterwards we do the general interpretation of options and there `RGB` is translated to color space #1.

The fix is to skip the PPD's options when doing the general interpretation, as the PPD has more precise information about the settings (the PPD is what the printer manufacturer or driver author provides to describe the properties of the printer and/or the driver). Ths wa already done in master, as the `Resolution` option of the Gutenprint driver caused a similar problem. Now I have ported this to 1.x.

**Comment by markwellis** [comment]
Great thanks very much

**Comment by tillkamppeter** [comment]
Fix is also included in today's 1.28.11 release of cups-filters.

**Comment by markwellis** [comment]
Excellent news, thanks very much :)
